charger/schneider.go

```diff
@@ -0,0 +1,222 @@
+	"github.com/evcc-io/evcc/util"
+	"github.com/evcc-io/evcc/util/modbus"
+	"github.com/evcc-io/evcc/util/sponsor"
+	"github.com/volkszaehler/mbmd/encoding"
+)
+
+// Schneider charger implementation
+type Schneider struct {
+	conn *modbus.Connection
+	curr uint16
+}
+
+const (
+	schneiderRegEvState       = 1
+	schneiderRegCurrents      = 2999
+	schneiderRegVoltages      = 3027
+	schneiderRegPower         = 3059
+	schneiderRegEnergy        = 3203
+	schneiderRegSetCommand    = 4001
+	schneiderRegSetPoint      = 4004
+	schneiderRegChargingTime  = 4007
+	schneiderRegSessionEnergy = 4012
+)
+
+func init() {
+	registry.Add("schneider", NewSchneiderFromConfig)
+}
+
+// https://download.schneider-electric.com/files?p_enDocType=Other+technical+guide&p_File_Name=GEX1969300-03.pdf&p_Doc_Ref=GEX1969300
+
+// NewSchneiderFromConfig creates a Schneider charger from generic config
+func NewSchneiderFromConfig(other map[string]interface{}) (api.Charger, error) {
+	cc := struct {
+		modbus.Settings `mapstructure:",squash"`
+		Timeout         time.Duration
+	}{
+		Settings: modbus.Settings{
+			ID: 255,
+		},
+	}
+
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	return NewSchneider(cc.URI, cc.Device, cc.Comset, cc.Baudrate, cc.ID, cc.Timeout)
+}
+
+// go:generate go run ../cmd/tools/decorate.go -f decorateSchneider -b *Schneider -r api.Charger -t "api.Meter,CurrentPower,func() (float64, error)" -t "api.PhaseCurrents,Currents,func() (float64, float64, float64, error)"
+
+// NewSchneider creates Schneider charger
+func NewSchneider(uri, device, comset string, baudrate int, slaveID uint8, timeout time.Duration) (api.Charger, error) {
+	conn, err := modbus.NewConnection(uri, device, comset, baudrate, modbus.Tcp, slaveID)
+	if err != nil {
+		return nil, err
+	}
+
+	if timeout > 0 {
+		conn.Timeout(timeout)
+	}
+
+	if !sponsor.IsAuthorized() {
+		return nil, api.ErrSponsorRequired
+	}
+
+	log := util.NewLogger("schneider")
+	conn.Logger(log.TRACE)
+
+	wb := &Schneider{
+		conn: conn,
+		curr: 6,
+	}
+
+	return wb, nil
+}
+
+// Status implements the api.Charger interface
+func (wb *Schneider) Status() (api.ChargeStatus, error) {
+	b, err := wb.conn.ReadHoldingRegisters(schneiderRegEvState, 1)
+	if err != nil {
+		return api.StatusNone, err
+	}
+
+	s := binary.BigEndian.Uint16(b)
+
+	switch s {
+	case 1, 2, 6:
+		return api.StatusA, nil
+	case 3, 4, 5, 7:
+		return api.StatusB, nil
+	case 8, 9:
+		return api.StatusC, nil
+	default:
+		return api.StatusNone, fmt.Errorf("invalid status: %d", s)
+	}
+}
+
+// Enabled implements the api.Charger interface
+func (wb *Schneider) Enabled() (bool, error) {
+	b, err := wb.conn.ReadHoldingRegisters(schneiderRegSetCommand, 1)
+	if err != nil {
+		return false, err
+	}
+
+	fmt.Println(binary.BigEndian.Uint16(b))
+
+	return binary.BigEndian.Uint16(b) > 0, nil
+}
+
+// Enable implements the api.Charger interface
+func (wb *Schneider) Enable(enable bool) error {
+	var u uint16
```

Comment:
```suggestion
	var u uint16 = 99
```

Hier bin ich mir unsicher ob man die 99 auch schreiben kann zum disablen